### PR TITLE
Annotate backups with kubernetes version information

### LINKED ISSUE #2429: Kubernetes minor version label causes errors when patching backups
**What steps did you take and what happened:**
[A clear and concise description of what the bug is, and what commands you ran.)
On a GKE cluster of v1.16.8-gke.8, taking a backup results in the backup failing to update its status with the following error on a perpetual loop:

```
ERRO[0019] Error in syncHandler, re-adding item to queue  controller=backup error="error updating Backup status to InProgress: error patching backup: Backup.velero.io \"nginx\" is invalid: metadata.labels: Invalid value: \"16+\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')" error.file="/Users/nrb/go/src/github.com/vmware-tanzu/velero/pkg/controller/backup_controller.go:303" error.function=github.com/vmware-tanzu/velero/pkg/controller.patchBackup key=velero/nginx logSource="pkg/controller/generic_controller.go:140"
```

When printing out the updated object, I see following patch being applied (bolding mine):

```go
&v1.Backup{TypeMeta:v1.TypeMeta{Kind:"Backup", APIVersion:"velero.io/v1"}, ObjectMeta:v1.ObjectMeta{Name:"nginx", GenerateName:"", Namespace:"velero", SelfLink:"/apis/velero.io/v1/namespaces/velero/backups/nginx", UID:"7a84d7c0-7903-463b-99f1-45f3a88ba021", ResourceVersion:"5515", Generation:1, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63722402918, loc:(*time.Location)(0x37f2900)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"velero.io/source-cluster-k8s-gitversion":"v1.16.8-gke.8", "velero.io/source-cluster-k8s-major-version":"1", "velero.io/source-cluster-k8s-minor-version":"16+", "velero.io/storage-location":"default"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.BackupSpec{IncludedNamespaces:[]string{"nginx-example"}, ExcludedNamespaces:[]string(nil), IncludedResources:[]string(nil), ExcludedResources:[]string(nil), LabelSelector:(*v1.LabelSelector)(nil), SnapshotVolumes:(*bool)(nil), TTL:v1.Duration{Duration:2592000000000000}, IncludeClusterResources:(*bool)(nil), Hooks:v1.BackupHooks{Resources:[]v1.BackupResourceHookSpec(nil)}, StorageLocation:"default", VolumeSnapshotLocations:[]string{"default"}}, Status:v1.BackupStatus{Version:1, Expiration:(*v1.Time)(0xc00000d3e0), Phase:"InProgress", ValidationErrors:[]string(nil), StartTimestamp:(*v1.Time)(0xc00000d4a0), CompletionTimestamp:(*v1.Time)(nil), VolumeSnapshotsAttempted:0, VolumeSnapshotsCompleted:0, Warnings:0, Errors:0}}
```

Specifically, this label: `"velero.io/source-cluster-k8s-minor-version":"16+"`

**What did you expect to happen:**
Labels applied by Velero should be valid & not cause errors with completing the backup.

- Velero version (use `velero version`): `master`, at [`fdba6a6c55e98d550c42e8533bfbd6d01a9a4604`](https://github.com/vmware-tanzu/velero/commit/fdba6a6c55e98d550c42e8533bfbd6d01a9a4604)
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:16:51Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.8-gke.8", GitCommit:"cbcdab5cd1710d6d7902daca273897d07a9e73a6", GitTreeState:"clean", BuildDate:"2020-04-01T10:06:36Z", GoVersion:"go1.13.8b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version: GKE
- Cloud provider or hardware configuration: GKE

### PR DESCRIPTION
**Pull request by nrb**
Fixes #2429 

Signed-off-by: Nolan Brubaker <brubakern@vmware.com>

### CODE REVIEW

**Comment by nrb** [comment]
Just double checked - this bug is present in master, but not in any released code yet.

**Comment by carlisia** [comment]
A test is failing.

**Comment by nrb** [comment]
Fixed the broken test by reverting [L333](https://github.com/vmware-tanzu/velero/pull/2431/files#diff-3569cd23a9de2def8b4cdf8b4d33fc6fL333) back to using the valid name function.

I think that decision may need to be revisited, though - this shouldn't be a backwards incompatible change (I *believe* it only affects new backups), but it does work on user-submitted values...so I'm not 100% sure.

I'm fine with this particular PR focusing on the Kubernetes version labels for the moment, though, as those were the ones that were failing for me when I was testing something completely unrelated.

**Comment by nrb** [comment]
Changing this PR up to write the version info as annotations instead of labels, since the regex is hard to get correct/maintain, and per @brito-rafa, we shouldn't need to query the API server for all backups made on a given version.

Also going to fix the newly broken tests.

File: pkg/apis/velero/v1/labels_annotations.go [hunk 1]
```diff
@@ -53,13 +53,13 @@ const (
 
 	// SourceClusterK8sVersionLabel is the label key used to identify the k8s
```
Comment (skriss) [inline pkg/apis/velero/v1/labels_annotations.go:54]: update godoc to match field names (for all three)

File: pkg/label/label.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ import (
 // If length exceeds, we trim the label name to contain only max allowed characters
 // Additionally, the last 6 characters of the label name are replaced by the first 6 characters of the sha256 of original label
 func GetValidName(label string) string {
+	// regex in here?
```
Comment (skriss) [inline pkg/label/label.go:32]: can drop this comment (I think)

**Comment by nrb** [comment]
@skriss 

>we've checked that the invalid GKE version string works as an annotation, right?

Yep, made sure that worked before pushing up the code, but forgot `make test`.

```
% v describe backup nginx
Name:         nginx
Namespace:    velero
Labels:       velero.io/storage-location=default
Annotations:  velero.io/source-cluster-k8s-gitversion=v1.16.8-gke.8
              velero.io/source-cluster-k8s-major-version=1
              velero.io/source-cluster-k8s-minor-version=16+

Phase:  Completed

Namespaces:
  Included:  nginx-example
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        <none>
  Cluster-scoped:  auto

Label selector:  <none>

Storage Location:  default

Snapshot PVs:  auto

TTL:  720h0m0s

Hooks:  <none>

Backup Format Version:  1

Started:    2020-04-14 16:31:24 -0400 EDT
Completed:  2020-04-14 16:31:31 -0400 EDT

Expiration:  2020-05-14 16:31:24 -0400 EDT

Persistent Volumes:  1 of 1 snapshots completed successfully (specify --details for more information)
```

**Comment by nrb** [comment]
@skriss I had started adding one originally, but opted against it since the bug itself wasn't in a released version of Velero.

However, since the labels are now moved to annotations, it may be worth maybe changing up the messages...

**Comment by skriss** [comment]
yeah, it'll be a little messy either way - can also just try to remember to hand-edit at release time, but we'll probably forget :)

**Comment by nrb** [comment]
I'll amend https://github.com/vmware-tanzu/velero/blob/master/changelogs/unreleased/2346-brito-rafa

**Pull request merged**